### DISCUSSION TITLE
Fix support for when Konsist is added along side the kotlin gradle plugin

### DISCUSSION THREAD

**Discussion by tevjef** [LemonAppDev/konsist#795] [Request a Feature]
### Tell us more.

A use case I currently have for this library is the leveraging it's query API to collect metrics about the code base. We collect metrics about ongoing migrations and would like to leverage the expressive API to make it easier for us to collect this data. Here is one example:

```kotlin
    internal fun countMockkTests(): Metric {
        val value = scope
            .files
            .withImport { it.hasNameContaining("io.mockk") }
            .size

        return Metric(
            name = "source.mockk.count",
            value = value,
        )
    }
```

One blocker we have is that this library introduces issues in our build when added along side the android and kotlin gradle plugin

```
    implementation(libs.konsist)
    implementation(libs.gradlePluginKotlin)
```

```
   > Failed to transform kotlinx-coroutines-core-jvm-1.7.3.jar to match attributes {artifactType=classpath-entry-snapshot, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for ClasspathEntrySnapshotTransform: /Users/tevjef/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-core-jvm/1.7.3/2b09627576f0989a436a00a4a54b55fa5026fb86/kotlinx-coroutines-core-jvm-1.7.3.jar.
         > class org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotExternalizer cannot be cast to class org.jetbrains.kotlin.com.intellij.util.io.DataExternalizer (org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotExternalizer and org.jetbrains.kotlin.com.intellij.util.io.DataExternalizer are in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader$InstrumentingVisitableURLClassLoader @3ac5b955)
   > Failed to transform kotlinx-coroutines-android-1.7.3.jar to match attributes {artifactType=classpath-entry-snapshot, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for ClasspathEntrySnapshotTransform: /Users/tevjef/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-android/1.7.3/38d9cad3a0b03a10453b56577984bdeb48edeed5/kotlinx-coroutines-android-1.7.3.jar.
         > class org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotExternalizer cannot be cast to class org.jetbrains.kotlin.com.intellij.util.io.DataExternalizer (org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotExternalizer and org.jetbrains.kotlin.com.intellij.util.io.DataExternalizer are in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader$InstrumentingVisitableURLClassLoader @3ac5b955)
   > Failed to transform redacted-compiler-plugin-annotations-jvm-1.6.1.jar to match attributes {artifactType=classpath-entry-snapshot, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for ClasspathEntrySnapshotTransform: /Users/tevjef/.gradle/caches/modules-2/files-2.1/dev.zacsweers.redacted/redacted-compiler-plugin-annotations-jvm/1.6.1/5891794be9c2d817faeded43bf4556b7435281f3/redacted-compiler-plugin-annotations-jvm-1.6.1.jar.
 
```

My hypothesis is that the kotlin-compiler dependency is introducing classes unto the classpath that clash with those added by the transitive kotlin compiler. 

I've tried replacing the `kotlin-compiler` dependency with `kotlin-compiler-embeddable` with little luck.

```kotlin
configurations.all {
    resolutionStrategy {
        dependencySubstitution {
            substitute(module("org.jetbrains.kotlin:kotlin-compiler")).using(
                module("org.jetbrains.kotlin:kotlin-compiler-embeddable:1.9.10")
            )
        }
    }
}

dependencies {
    // To get org.intellij dependencies that are not in kotlin-compiler-embeddable
    implementation("com.android.tools.external.com-intellij:intellij-core:31.1.4")

    implementation(libs.konsist)
    implementation(libs.gradlePluginKotlin)
}
```

After substituting the dependency I now get this: 
```
Caused by: java.lang.ClassCastException: class org.jetbrains.kotlin.idea.KotlinFileType cannot be cast to class com.intellij.openapi.fileTypes.FileType (org.jetbrains.kotlin.idea.KotlinFileType and com.intellij.openapi.fileTypes.FileType are in unnamed module of loader 'app')
	at com.lemonappdev.konsist.core.util.KotlinFileParser.getKtFile(KotlinFileParser.kt:44)
	... 100 more
```

I understand this use case is not the primary use-case of the library, but I see a lot of value the in the query API as a tool build other tools.

**Comment by igorwojda** [comment]
@tevjef [Konsist v0.15.0](https://github.com/LemonAppDev/konsist/releases/tag/v0.15.0) is using `kotlin-compiler-embeddable`. Please give ta try.

**Comment by Jolanrensen** [comment]
Actually, I get the same exception with konsist 0.15.1 when running it from a gradle plugin, even when using shadow() on konsist, gradleApi() and gradleKotlinDsl():

```
Caused by: java.lang.ClassCastException: class org.jetbrains.kotlin.idea.KotlinFileType cannot be cast to class org.jetbrains.kotlin.com.intellij.openapi.fileTypes.FileType (org.jetbrains.kotlin.idea.KotlinFileType and org.jetbrains.kotlin.com.intellij.openapi.fileTypes.FileType are in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @590e1089)
	at com.lemonappdev.konsist.core.util.KotlinFileParser.getKtFile(KotlinFileParser.kt:45)
```

Any ideas?

**Reply by igorwojda** [reply]
@Jolanrensen Any chance you can share a minimal project to hep us reprode this issue?

**Reply by Jolanrensen** [reply]
I'll try, currently my project is very large haha :)

**Reply by igorwojda** [reply]
I thought this may be the case, but I imagine it's build configuration (dependencies) that are causing this. 
It may be best to remove bits of the project until issue is reproducible (together with all essential Kotlin files that shouldn't be shared).

If you have sensitive project (and can share it, or bits of it) consider sharing it with me via DM on Kotlin Sack https://kotlinlang.slack.com/archives/D2T3KL43X

**Reply by Jolanrensen** [reply]
I can actually share it :) it's open-source anyways.

I made a branch [here](https://github.com/Jolanrensen/docProcessorGradlePlugin/tree/konsist-test) which contains the problem. I suspected dokka dependencies might interfere, so I disabled those but the problem is still there :/.

The relevant gradle file with dependencies can be found [here](https://github.com/Jolanrensen/docProcessorGradlePlugin/blob/konsist-test/doc-processor-gradle-plugin/build.gradle.kts).
[This](https://github.com/Jolanrensen/docProcessorGradlePlugin/blob/d9f8f32451b4591665e7634292528dcfe7e56899/doc-processor-gradle-plugin/src/main/kotlin/nl/jolanrensen/docProcessor/ProcessDocsAction.kt#L48) is where I call into Konsist from the gradle plugin. You can put a breakpoint here to debug.

To run a functional test and observe the exception, you'll first need to `gradlew publishToMavenLocal` and then you can run a test like [this](https://github.com/Jolanrensen/docProcessorGradlePlugin/blob/d9f8f32451b4591665e7634292528dcfe7e56899/doc-processor-gradle-plugin/src/functionalTest/kotlin/nl/jolanrensen/docProcessor/TestInclude.kt#L20).

Hopefully it's not too difficult to reproduce it :/

**Reply by Jolanrensen** [reply]
Acutally, I found another place where I inject `kotlin-compiler` and `dokka` dependencies into the user's classpath. Removing that removes the exception :) 

So the key takeaway is to be careful no other version of the kotlin (embeddable) compiler exists on the classpath.

**Reply by igorwojda** [reply]
Great discovery thx for sharing
